src/Hyperion/Extensions/TypeEx.cs

```diff
@@ -208,16 +226,16 @@ public static Type LoadTypeByName(string name, bool disallowUnsafeTypes, ITypeFi
                 // i.e. if there are different version available in GAC and locally
                 var typename = ToQualifiedAssemblyName(name, ignoreAssemblyVersion: false);
                 var type = Type.GetType(typename, true);
-                if (UnsafeInheritanceCheck(type))
+                if (disallowUnsafeTypes && UnsafeInheritanceCheck(type))
                     throw new EvilDeserializationException(
                         "Unsafe Type Deserialization Detected!", name);
                 return type;
             }
             catch (IOException)
             {
                 var typename = ToQualifiedAssemblyName(name, ignoreAssemblyVersion: true);
-                var type =  Type.GetType(typename, true);
-                if (UnsafeInheritanceCheck(type))
+                var type = Type.GetType(typename, true);
+                if (disallowUnsafeTypes && UnsafeInheritanceCheck(type))
```

Comment:
This is the bug, disallowed types are checked several times, this line need to be bypassed if unsafe types are allowed